solidago/src/solidago/mehestan/individual/continuous_bradly_terry.py

```diff
@@ -0,0 +1,156 @@
+        """
+        self.alpha = alpha
+        self.r_max = r_max
+
+    def coordinate_descent(self, coord_to_subset, initial_scores):
+        n_alternatives = len(coord_to_subset)
+        unchanged = set()
+        to_pick = []
+
+        def pick_next_coordinate():
+            nonlocal to_pick
+            if len(to_pick) == 0:
+                to_pick = list(range(n_alternatives))
+                random.shuffle(to_pick)
+            return to_pick.pop()
+
+        theta = initial_scores
+        while len(unchanged) < n_alternatives:
+            coord = pick_next_coordinate()
+            if coord in unchanged:
+                continue
+            indices, r_ab = coord_to_subset[coord]
+            old_theta_a = theta[coord]
+            theta_b = theta[indices]
+            new_theta_a = coordinate_optimize(
+                r_ab, theta_b, precision=EPSILON / 10, alpha=self.alpha
+            )
+            theta[coord] = new_theta_a
+            if abs(new_theta_a - old_theta_a) < EPSILON:
+                unchanged.add(coord)
+            else:
+                unchanged.clear()
+        return theta
+
+    def compute_individual_scores(self, scores: pd.DataFrame, initial_entity_scores=None):
+        scores = scores[["entity_a", "entity_b", "score"]]
+        scores_sym = pd.concat(
+            [
+                scores,
+                pd.DataFrame(
+                    {
+                        "entity_a": scores.entity_b,
+                        "entity_b": scores.entity_a,
+                        "score": -1 * scores.score,
+                    }
+                ),
+            ]
+        )
+        r = scores_sym.pivot(index="entity_a", columns="entity_b", values="score") / self.r_max
+        r_values = r.to_numpy()
+        n_entities = len(r_values)
+        coord_to_subset = {}
+        for coord in range(n_entities):
+            r_ab = r_values[coord, :]
+            indices = (~np.isnan(r_ab)).nonzero()
+            coord_to_subset[coord] = (indices, r_ab[indices])
+
+        if initial_entity_scores is None:
+            initial_scores = np.zeros(n_entities)
+        else:
+            initial_scores = pd.Series(initial_entity_scores, index=r.index)
+            initial_scores.fillna(0.0, inplace=True)
+            initial_scores = initial_scores.to_numpy()
+        theta_star_numpy = self.coordinate_descent(coord_to_subset, initial_scores=initial_scores)
+        delta_star_numpy = np.zeros(len(theta_star_numpy))
+        for idx in range(len(theta_star_numpy)):
+            indices, _r_ab = coord_to_subset[idx]
+            delta_star_numpy[idx] = Delta_theta(theta_star_numpy[indices])
+
+        result = pd.DataFrame(
+            {
+                "raw_score": theta_star_numpy,
+                "raw_uncertainty": delta_star_numpy,
+            },
+            index=r.index,
+        )
+        result.index.name = "entity_id"
+        return result
+
+    def get_metadata(self) -> dict:
```

Comment:
I'm very glad we created the Solidago library :)

We will regularly encounter the question: "Is it solidago's or tournesol's responsibility?"

For the case of this method `get_metadata` I would slightly (not a strong opinion) be on the side of tournesol's responsibility. Typically we want to only make changes to Solidago when we implement a new algorithm, not when we add a feature to Tournesol.

Reply:
Even if Tournesol is responsable for building the metadata attached to the public dataset, I think that it makes sense for each algorithm to expose its entire parameter set. Otherwise Tournesol's backend would need to be know about solidago implementation details to write the public dataset, and it would make switching between algorithms more cumbersome.

Or do you have another alternative in mind?